src/custom-elements/src/components/number-input/number-input.tsx

```diff
@@ -84,114 +112,98 @@ export class NumberInput {
     CallbackQueue.queueProcess(this._didUpdateCallBackOnceQueue);
   }
 
-  static isNumeric(n: string | number) {
-    return (
-      (typeof n === 'number' || typeof n === 'string') &&
-      !isNaN(+n - parseFloat(`${n}`))
-    );
-  }
+  handleChange(ev: Event) {
+    ev.stopPropagation();
 
-  _setNewValue(newValue: string) {
     if (!this.preventValueMutation) {
-      this.value = newValue;
       this._didUpdateCallBackOnceQueue.push(() => {
-        this.chiChange.emit(newValue);
+        this.chiChange.emit(this.value);
       });
     } else {
-      this.chiChange.emit(newValue);
+      this.chiChange.emit(this.value);
     }
   }
 
-  handleChange(ev: Event) {
+  handleInput(ev: Event) {
+    const input = ev.target as HTMLInputElement;
+
     ev.stopPropagation();
 
-    let newValue = !!ev.target
-      ? (ev.target as HTMLInputElement).value.toString()
-      : null;
-    newValue = NumberInput.isNumeric(newValue) ? newValue : null;
-
-    if (newValue !== null) {
-      let steppedValue =
-        Math.round(10000 * ((+newValue - this.initialValue) / this.step)) /
-        10000;
-      const steppedMax =
-        Math.round(
-          10000 * Math.floor((+this.max - this.initialValue) / this.step)
-        ) / 10000;
-      const steppedMin =
-        Math.round(
-          10000 * Math.ceil((+this.min - this.initialValue) / this.step)
-        ) / 10000;
-
-      steppedValue = Math.max(Math.min(steppedValue, steppedMax), steppedMin);
-      newValue = (
-        Math.round(steppedValue) * this.step +
-        +this.initialValue
-      ).toString();
-    }
+    this.value = input.value;
+    this.chiInput.emit(this.value);
 
-    if (newValue !== this.value) {
-      this._setNewValue(newValue);
-    } else {
-      if ((ev.target as HTMLInputElement).value.toString() !== this.value) {
-        (ev.target as HTMLInputElement).value = this.value;
-      }
-    }
+    if (!input.validity.valid) this.chiNumberInvalid.emit();
   }
 
   private increment() {
-    const step = this.step;
-    const newValue = Math.round((+this.value + step) * 10000) / 10000;
+    const input = this.el.id
+      ? (document.getElementById(`${this.el.id}-control`) as HTMLInputElement)
+      : (this.el.children[0].children[0] as HTMLInputElement);
+
+    input.stepUp();
+    this.value = this.setNewValue(input.value);
```

Comment:
Why is not this part of the `if (input.valueAsNumber <= this.max)` condition?
Please, check if it updates the input value with a new number in case it goes over the max.